src/clj/puppetlabs/trapperkeeper/services/watcher/filesystem_watch_core.clj

```diff
@@ -81,42 +92,58 @@
    watcher :- (schema/protocol Watcher)]
   (let [dir-create? (fn [event]
                       (and (= :create (:type event))
-                           (fs/directory? (:path event))))]
+                           (fs/directory? (:full-path event))))]
     (DirWatchUtils/registerRecursive (:watch-service watcher)
                                      (->> events
                                           (filter dir-create?)
-                                          (map #(.toPath (:path %)))))))
+                                          (map #(.toPath (:full-path %)))))))
+
+(schema/defn get-event-maps-from-key :- [Event]
+  [watch-key :- WatchKey]
+  (let [events (.pollEvents watch-key)]
+    (map #(clojurize % (.watchable watch-key)) events)))
 
-(schema/defn retrieve-events
-  :- [(schema/one WatchKey "key") (schema/one [WatchEvent] "events")]
-  "Blocks until an event the watcher is concerned with has occured.
-  Returns the native WatchKey and WatchEvents"
+(schema/defn retrieve-events :- [Event]
+  "Blocks until an event the watcher is concerned with has occured. Will then
+  poll for a new event, waiting at least `window-min` for a new event to
+  occur. Will continue polling for as long as there are new events that occur
+  within `window-min`, or the `window-max` time limit has been exceeded."
   [watcher :- (schema/protocol Watcher)]
   (let [watch-key (.take (:watch-service watcher))
-        events (.pollEvents watch-key)]
+        events (get-event-maps-from-key watch-key)
```

Comment:
Oh, I just now saw the discussion at https://github.com/puppetlabs/trapperkeeper-filesystem-watcher/pull/8#discussion_r73751528

I still think it would be good to check if `.pollEvents` returned an empty list and, if so, not loop.